### PR TITLE
Subscription EndpointAddressResolverMetadata

### LINKED ISSUE #514: Subscription metadata
Subscriptions model should allow for storing additional metadata.

### PR DESCRIPTION
**Pull request by dankraw**
Introducing EndpointAddressResolverMetadata. `Map` implementation.
resolves #514

### CODE REVIEW

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        return entries.stream().filter(e -> e.getKey().equals(key)).findFirst();
+    }
+
+    public List<MetadataEntry> getEntries() {
+        return entries;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java]: \[Checkstyle\] ERROR: 'if' construct must use '\{\}'s\.

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        if (o == null || getClass() != o.getClass()) return false;
```
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java]: \[Checkstyle\] ERROR: 'if' construct must use '\{\}'s\.

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java (hunk 1 above)
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java]: \[PMD\] INFO: Avoid using if statements without curly braces

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java (hunk 2 above)
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java]: \[PMD\] INFO: Avoid using if statements without curly braces

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/MetadataEntry.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        return key;
+    }
+
+    public Object getValue() {
+        return value;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/MetadataEntry.java]: \[Checkstyle\] ERROR: 'if' construct must use '\{\}'s\.

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/MetadataEntry.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        if (o == null || getClass() != o.getClass()) return false;
```
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/MetadataEntry.java]: \[Checkstyle\] ERROR: 'if' construct must use '\{\}'s\.

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java [hunk 5]
```diff
@@ -129,9 +134,10 @@ public static Subscription createBatchSubscription(TopicName topicName,
                                                        MonitoringDetails monitoringDetails,
                                                        ContentType contentType,
                                                        List<MessageFilterSpecification> filters,
-                                                       List<Header> headers) {
+                                                       List<Header> headers,
+                                                       EndpointAddressResolverMetadata endpointAddressResolverMetadata) {
         return new Subscription(topicName, name, endpoint, state, description, subscriptionPolicy, trackingEnabled, supportTeam,
-                contact, monitoringDetails, contentType, DeliveryType.BATCH, filters, SubscriptionMode.ANYCAST, headers);
+                contact, monitoringDetails, contentType, DeliveryType.BATCH, filters, SubscriptionMode.ANYCAST, headers, endpointAddressResolverMetadata);
```
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 154\)\.

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java [hunk 6]
```diff
@@ -149,7 +155,8 @@ public static Subscription create(@JsonProperty("topicName") String topicName,
                                       @JsonProperty("deliveryType") DeliveryType deliveryType,
                                       @JsonProperty("filters") List<MessageFilterSpecification> filters,
                                       @JsonProperty("mode") SubscriptionMode mode,
-                                      @JsonProperty("headers") List<Header> headers) {
+                                      @JsonProperty("headers") List<Header> headers,
+                                      @JsonProperty("endpointAddressResolverMetadata") EndpointAddressResolverMetadata endpointAddressResolverMetadata) {
```
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 153\)\.

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java [hunk 7]
```diff
@@ -171,14 +178,15 @@ public static Subscription create(@JsonProperty("topicName") String topicName,
...
+                headers == null ? Collections.emptyList() : headers,
+                endpointAddressResolverMetadata == null ? EndpointAddressResolverMetadata.empty() : endpointAddressResolverMetadata
         );
     }
 
     @Override
     public int hashCode() {
         return Objects.hash(endpoint, topicName, name, description, serialSubscriptionPolicy, batchSubscriptionPolicy,
-                trackingEnabled, supportTeam, contact, monitoringDetails, contentType, filters, mode, headers);
+                trackingEnabled, supportTeam, contact, monitoringDetails, contentType, filters, mode, headers, endpointAddressResolverMetadata);
```
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 144\)\.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/ConsumerMessageSender.java [hunk 8]
```diff
@@ -91,9 +91,10 @@ public void sendMessage(final Message message) {
     public synchronized void updateSubscription(Subscription newSubscription) {
         boolean endpointUpdated = !this.subscription.getEndpoint().equals(newSubscription.getEndpoint());
         boolean subscriptionPolicyUpdated = !Objects.equals(this.subscription.getSerialSubscriptionPolicy(), newSubscription.getSerialSubscriptionPolicy());
+        boolean endpointAddressResolverMetadataChanged = !Objects.equals(this.subscription.getEndpointAddressResolverMetadata(), newSubscription.getEndpointAddressResolverMetadata());
```
Comment (sputnikci) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/ConsumerMessageSender.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 183\)\.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ApacheHttpClientMessageBatchSender.java [hunk 9]
```diff
@@ -36,9 +37,9 @@ public ApacheHttpClientMessageBatchSender(int connectionTimeout, int socketTimeo
     }
 
     @Override
-    public MessageSendingResult send(MessageBatch batch, EndpointAddress address, int requestTimeout) {
+    public MessageSendingResult send(MessageBatch batch, EndpointAddress address, EndpointAddressResolverMetadata metadata, int requestTimeout) {
```
Comment (sputnikci) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ApacheHttpClientMessageBatchSender.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 145\)\.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/EndpointAddressResolver.java [hunk 10]
```diff
@@ -10,16 +11,16 @@
 
 public interface EndpointAddressResolver {
 
-    default URI resolve(EndpointAddress address, Message message) throws EndpointAddressResolutionException {
+    default URI resolve(EndpointAddress address, Message message, EndpointAddressResolverMetadata metadata) throws EndpointAddressResolutionException {
```
Comment (sputnikci) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/EndpointAddressResolver.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 151\)\.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/EndpointAddressResolver.java [hunk 11]
```diff
@@ -10,16 +11,16 @@
...
         return resolve(address);
     }
 
-    default URI resolve(EndpointAddress address, MessageBatch batch) throws EndpointAddressResolutionException {
+    default URI resolve(EndpointAddress address, MessageBatch batch, EndpointAddressResolverMetadata metadata) throws EndpointAddressResolutionException {
```
Comment (sputnikci) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/EndpointAddressResolver.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 154\)\.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/EndpointAddressResolver.java [hunk 12]
```diff
@@ -10,16 +11,16 @@
...
         return resolve(address);
     }
 
-    default List<URI> resolveAll(EndpointAddress address, Message message) throws EndpointAddressResolutionException {
-        return Collections.singletonList(resolve(address, message));
+    default List<URI> resolveAll(EndpointAddress address, Message message, EndpointAddressResolverMetadata metadata) throws EndpointAddressResolutionException {
```
Comment (sputnikci) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/EndpointAddressResolver.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 160\)\.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/InterpolatingEndpointAddressResolver.java [hunk 13]
```diff
@@ -18,7 +19,7 @@ public InterpolatingEndpointAddressResolver(UriInterpolator interpolator) {
     }
 
     @Override
-    public URI resolve(EndpointAddress address, Message message) throws EndpointAddressResolutionException {
+    public URI resolve(EndpointAddress address, Message message, EndpointAddressResolverMetadata metadata) throws EndpointAddressResolutionException {
```
Comment (sputnikci) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/InterpolatingEndpointAddressResolver.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 150\)\.

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+    @NotNull
+    private Map<String, Object> entries;
+
+    @JsonCreator
+    public EndpointAddressResolverMetadata(Map<String, Object> entries) {
+        this.entries = ImmutableMap.copyOf(entries);
+    }
+
+    public Optional<Object> get(String key) {
```
Comment (wojtkiewicz) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java]: you could add ```getOrDefault```

Comment (wojtkiewicz) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java]: and ```getStringOrDefault``` or ```getIntOrDefault```

**Comment by wojtkiewicz** [comment]
👍

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java (hunk 14 above)
Comment (dankraw) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/EndpointAddressResolverMetadata.java]: Added a generic
```java
public <T> T getOrDefault(String key, T defaultValue) {
    return (T) entries.getOrDefault(key, defaultValue);
}
```

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java [hunk 15]
```diff
@@ -129,27 +134,31 @@ public static Subscription createBatchSubscription(TopicName topicName,
...
-                                      @JsonProperty("trackingEnabled") boolean trackingEnabled,
-                                      @JsonProperty("supportTeam") String supportTeam,
-                                      @JsonProperty("contact") String contact,
-                                      @JsonProperty("monitoringDetails") MonitoringDetails monitoringDetails,
-                                      @JsonProperty("contentType") ContentType contentType,
-                                      @JsonProperty("deliveryType") DeliveryType deliveryType,
-                                      @JsonProperty("filters") List<MessageFilterSpecification> filters,
-                                      @JsonProperty("mode") SubscriptionMode mode,
-                                      @JsonProperty("headers") List<Header> headers) {
+    public static Subscription create(
```
Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java]: \[Checkstyle\] ERROR: More than 12 parameters \(found 16\)\.

Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java]: \[Checkstyle\] ERROR: More than 7 parameters \(found 16\)\.

Comment (sputnikci) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/Subscription.java]: \[Checkstyle\] ERROR: More than 7 parameters \(found 16\)\.

**Comment by pbetkier** [comment]
👍

**Pull request merged**